--- ./src/google/protobuf/stubs/strutil.h.orig	2023-04-05 11:43:33.660651000 +0100
+++ ./src/google/protobuf/stubs/strutil.h	2023-04-05 13:44:25.463427000 +0100
@@ -604,7 +604,11 @@
     // their unsigned counterparts.
 #ifdef LANG_CXX11
     static_assert(
+#if defined(__CHERI_PURE_CAPABILITY__)
+        sizeof(v) == 1 || sizeof(v) == 2 || sizeof(v) == 4 || sizeof(v) == 8 || sizeof(v) == 16,
+#else
         sizeof(v) == 1 || sizeof(v) == 2 || sizeof(v) == 4 || sizeof(v) == 8,
+#endif
         "Unknown integer type");
 #endif
     value = sizeof(v) == 1 ? static_cast<uint8>(v)
--- ./src/google/protobuf/generated_message_tctable_impl.h.orig	2023-04-05 11:24:06.649855000 +0100
+++ ./src/google/protobuf/generated_message_tctable_impl.h	2023-04-05 11:25:10.911680000 +0100
@@ -252,7 +252,7 @@
 [[noreturn]]
 #endif
 void AlignFail(uintptr_t address) {
-  GOOGLE_LOG(FATAL) << "Unaligned (" << align << ") access at " << address;
+  GOOGLE_LOG(FATAL) << "Unaligned (" << align << ") access at " << (size_t) address;
 }
 
 extern template void AlignFail<4>(uintptr_t);
--- ./src/google/protobuf/generated_message_tctable_decl.h.orig	2023-04-05 11:32:54.291557000 +0100
+++ ./src/google/protobuf/generated_message_tctable_decl.h	2023-04-05 14:05:40.615750000 +0100
@@ -121,7 +121,11 @@
 #endif
 
 // Base class for message-level table with info for the tail-call parser.
+#if defined(__CHERI_PURE_CAPABILITY__)
+struct alignas(max_align_t) TcParseTableBase {
+#else
 struct alignas(uint64_t) TcParseTableBase {
+#endif
   // Common attributes for message layout:
   uint16_t has_bits_offset;
   uint16_t extension_offset;
@@ -240,9 +244,17 @@
 #pragma warning(pop)
 #endif
 
+#if defined(__CHERI_PURE_CAPABILITY__)
+static_assert(sizeof(TcParseTableBase::FastFieldEntry) <= 32,
+#else
 static_assert(sizeof(TcParseTableBase::FastFieldEntry) <= 16,
+#endif
               "Fast field entry is too big.");
+#if defined(__CHERI_PURE_CAPABILITY__)
+static_assert(sizeof(TcParseTableBase::FieldEntry) <= 32,
+#else
 static_assert(sizeof(TcParseTableBase::FieldEntry) <= 16,
+#endif
               "Field entry is too big.");
 
 template <size_t kFastTableSizeLog2, size_t kNumFieldEntries = 0,
--- ./src/google/protobuf/map.h.orig	2022-04-21 20:47:49.000000000 +0100
+++ ./src/google/protobuf/map.h	2023-04-05 14:28:49.627957000 +0100
@@ -118,10 +118,14 @@
   MapAllocator(const MapAllocator<X>& allocator)  // NOLINT(runtime/explicit)
       : arena_(allocator.arena()) {}
 
+#if defined(__CHERI_PURE_CAPABILITY__)
+  static_assert(alignof(value_type) <= sizeof(max_align_t), "");
+#else
   // MapAllocator does not support alignments beyond 8. Technically we should
   // support up to std::max_align_t, but this fails with ubsan and tcmalloc
   // debug allocation logic which assume 8 as default alignment.
   static_assert(alignof(value_type) <= 8, "");
+#endif
 
   pointer allocate(size_type n, const void* /* hint */ = nullptr) {
     // If arena is not given, malloc needs to be called which doesn't
--- ./src/google/protobuf/descriptor.cc.orig	2022-04-21 20:47:49.000000000 +0100
+++ ./src/google/protobuf/descriptor.cc	2023-04-05 14:50:14.020313000 +0100
@@ -368,8 +368,12 @@
     // We can't call PlanArray after FinalizePlanning has been called.
     GOOGLE_CHECK(!has_allocated());
     if (std::is_trivially_destructible<U>::value) {
+#if defined(__CHERI_PURE_CAPABILITY__)
+      static_assert(alignof(U) <= sizeof(max_align_t), "");
+#else
       // Trivial types are aligned to 8 bytes.
       static_assert(alignof(U) <= 8, "");
+#endif
       total_.template Get<char>() += RoundUpTo<8>(array_size * sizeof(U));
     } else {
       // Since we can't use `if constexpr`, just make the expression compile
@@ -1806,7 +1810,11 @@
 template <typename Type>
 Type* DescriptorPool::Tables::Allocate() {
   static_assert(std::is_trivially_destructible<Type>::value, "");
+#if defined(__CHERI_PURE_CAPABILITY__)
+  static_assert(alignof(Type) <= sizeof(max_align_t), "");
+#else
   static_assert(alignof(Type) <= 8, "");
+#endif
   return ::new (AllocateBytes(sizeof(Type))) Type{};
 }
 
--- src/google/protobuf/metadata_lite.h.orig	2022-04-21 20:47:49.000000000 +0100
+++ src/google/protobuf/metadata_lite.h	2023-04-05 15:44:42.885407000 +0100
@@ -183,11 +183,19 @@
   intptr_t ptr_;
 
   // Tagged pointer implementation.
+#if defined(__CHERI_PURE_CAPABILITY__)
+  static constexpr ptraddr_t kUnknownFieldsTagMask = 1;
+  static constexpr ptraddr_t kMessageOwnedArenaTagMask = 2;
+  static constexpr ptraddr_t kPtrTagMask =
+      kUnknownFieldsTagMask | kMessageOwnedArenaTagMask;
+  static constexpr ptraddr_t kPtrValueMask = ~kPtrTagMask;
+#else
   static constexpr intptr_t kUnknownFieldsTagMask = 1;
   static constexpr intptr_t kMessageOwnedArenaTagMask = 2;
   static constexpr intptr_t kPtrTagMask =
       kUnknownFieldsTagMask | kMessageOwnedArenaTagMask;
   static constexpr intptr_t kPtrValueMask = ~kPtrTagMask;
+#endif
 
   // Accessors for pointer tag and pointer value.
   PROTOBUF_ALWAYS_INLINE bool HasUnknownFieldsTag() const {
@@ -218,7 +226,11 @@
       // Subtle: we want to preserve the message-owned arena flag, while at the
       // same time replacing the pointer to Container<T> with a pointer to the
       // arena.
+#if defined(__CHERI_PURE_CAPABILITY__)
+      ptraddr_t message_owned_arena_tag = ptr_ & kMessageOwnedArenaTagMask;
+#else
       intptr_t message_owned_arena_tag = ptr_ & kMessageOwnedArenaTagMask;
+#endif
       ptr_ = reinterpret_cast<intptr_t>(a) | message_owned_arena_tag;
       return a;
     } else {
--- src/google/protobuf/arena_impl.h.orig	2022-04-21 20:47:48.000000000 +0100
+++ src/google/protobuf/arena_impl.h	2023-04-05 15:55:55.847283000 +0100
@@ -121,7 +121,11 @@
       : policy_(reinterpret_cast<uintptr_t>(policy)) {}
 
   void set_policy(AllocationPolicy* policy) {
+#if defined(__CHERI_PURE_CAPABILITY__)
+    ptraddr_t bits = policy_ & kTagsMask;
+#else
     auto bits = policy_ & kTagsMask;
+#endif
     policy_ = reinterpret_cast<uintptr_t>(policy) | bits;
   }
 
@@ -163,19 +167,36 @@
     kRecordAllocs = 2,
   };
 
+#if defined(__CHERI_PURE_CAPABILITY__)
+  static constexpr ptraddr_t kTagsMask = 7;
+  static constexpr ptraddr_t kPtrMask = ~kTagsMask;
+#else
   static constexpr uintptr_t kTagsMask = 7;
   static constexpr uintptr_t kPtrMask = ~kTagsMask;
+#endif
 
   template <uintptr_t kMask>
   uintptr_t get_mask() const {
+#if defined(__CHERI_PURE_CAPABILITY__)
+    return policy_ & (ptraddr_t) kMask;
+#else
     return policy_ & kMask;
+#endif
   }
   template <uintptr_t kMask>
   void set_mask(bool v) {
     if (v) {
+#if defined(__CHERI_PURE_CAPABILITY__)
+      policy_ |= (ptraddr_t) kMask;
+#else
       policy_ |= kMask;
+#endif
     } else {
+#if defined(__CHERI_PURE_CAPABILITY__)
+      policy_ &= ~(ptraddr_t) kMask;
+#else
       policy_ &= ~kMask;
+#endif
     }
   }
   uintptr_t policy_;
