--- data/skeletons/yacc.c.orig	2021-09-12 10:42:54.000000000 +0100
+++ data/skeletons/yacc.c	2026-07-13 12:51:07.760818000 +0100
@@ -610,6 +610,12 @@
 
 # define YYCOPY_NEEDED 1
 
+# include <sys/cdefs.h>
+
+# ifndef __unbounded_addressof
+#  define __unbounded_addressof(x) (&(x))
+# endif
+
 /* Relocate STACK from its old location to the new one.  The
    local variables YYSIZE and YYSTACKSIZE give the old and new number of
    elements in the stack, and YYPTR gives the new location of the
@@ -619,8 +625,8 @@
     do                                                                  \
       {                                                                 \
         YYPTRDIFF_T yynewbytes;                                         \
-        YYCOPY (&yyptr->Stack_alloc, Stack, yysize);                    \
-        Stack = &yyptr->Stack_alloc;                                    \
+        YYCOPY (__unbounded_addressof(yyptr->Stack_alloc), Stack, yysize); \
+        Stack = __unbounded_addressof(yyptr->Stack_alloc);              \
         yynewbytes = yystacksize * YYSIZEOF (*Stack) + YYSTACK_GAP_MAXIMUM; \
         yyptr += yynewbytes / YYSIZEOF (*yyptr);                        \
       }                                                                 \
